src/Microsoft.ML.CpuMath/FactorizationMachine/FmaIntrinsics.cs

```diff
@@ -0,0 +1,188 @@
+            float* linearAccumulatedSquaredGrads, float* latentAccumulatedSquaredGrads)
+        {
+            int m = fieldCount;
+            int d = latentDim;
+            int c = count;
+            int* pf = fieldIndices;
+            int* pi = featureIndices;
+            float* px = featureValues;
+            float* pq = latentSum;
+            float* pw = linearWeights;
+            float* pv = latentWeights;
+            float* phw = linearAccumulatedSquaredGrads;
+            float* phv = latentAccumulatedSquaredGrads;
+
+            Vector256<float> wei = Vector256.Create(weight);
+            Vector256<float> s= Vector256.Create(slope);
+            Vector256<float> lr = Vector256.Create(learningRate);
+            Vector256<float> lambdav = Vector256.Create(lambdaLatent);
+
+            for (int i = 0; i < count; i++)
+            {
+                int f = pf[i];
+                int j = pi[i];
+
+                // Calculate gradient of linear term w_j.
+                float g = weight * (lambdaLinear * pw[j] + slope * px[i]);
+
+                // Accumulate the gradient of the linear term.
+                phw[j] += g * g;
+
+                // Perform ADAGRAD update rule to adjust linear term.
+                pw[j] -= learningRate / MathF.Sqrt(phw[j]) * g;
+
+                // Update latent term, v_j,f', f'=1,...,m.
+                Vector256<float> x = Avx.BroadcastScalarToVector256(px + i);
+
+                for (int fprime = 0; fprime < m; fprime++)
+                {
+                    float* vjfprime = pv + j * m * d + fprime * d;
+                    float* hvjfprime = phv + j * m * d + fprime * d;
+                    float* qfprimef = pq + fprime * m * d + f * d;
+                    Vector256<float> sx = Avx.Multiply(s, x);
+
+                    for (int k = 0; k + 8 <= d; k += 8)
+                    {
+                        Vector256<float> v = Avx.LoadVector256(vjfprime + k);
+                        Vector256<float> q = Avx.LoadVector256(qfprimef + k);
+
+                        // Calculate L2-norm regularization's gradient.
+                        Vector256<float> gLatent = Avx.Multiply(lambdav, v);
+
+                        // Calculate loss function's gradient.
+                        if (fprime != f)
+                            gLatent = Fma.MultiplyAdd(sx, q, gLatent);
+                        else
+                            gLatent = Fma.MultiplyAdd(sx, Fma.MultiplyAddNegated(v, x, q), gLatent);
+                        gLatent = Avx.Multiply(wei, gLatent);
+
+                        // Accumulate the gradient of latent vectors.
+                        Vector256<float> h = Fma.MultiplyAdd(gLatent, gLatent, Avx.LoadVector256(hvjfprime + k));
+
+                        // Perform ADAGRAD update rule to adjust latent vector.
+                        v = Fma.MultiplyAddNegated(lr, Avx.Multiply(Avx.ReciprocalSqrt(h), gLatent), v);
```

Comment:
For my knowledge (I'm not a SIMD expert), what is the difference between `MultiplyAddNegated` and `MultiplySubtract`?

Reply:
The main difference is the order in which the vectors get multiplied and added/subtracted.

MultiplyAddNegated instruction translates to `__m128 _mm_fnmadd_ps (__m128 a, __m128 b, __m128 c)` instruction. The  result is computed as: -(a*b) + c

whereas `MultiplySubtract` instruction translates to `__m128 _mm_fmsub_ps (__m128 a, __m128 b, __m128 c) `.
The result for this is computed as: (a*b) - c

[More info](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#expand=3847,4166,2697,2709&techs=FMA&text=%2520_mm_fnmadd_ps)